File: solutions/java/list-ops/1/src/main/java/ListOps.java
import java.util.ArrayList;
import java.util.Collection;
import java.util.Collections;
import java.util.List;
import java.util.function.BiFunction;
import java.util.function.Function;
import java.util.function.Predicate;
import java.util.stream.Collectors;
import java.util.stream.IntStream;
import java.util.stream.Stream;

class ListOps {

    static <T> List<T> append(List<T> list1, List<T> list2) {
        return Stream.of(list1, list2).flatMap(Collection::stream).collect(Collectors.toList());
    }

    static <T> List<T> concat(List<List<T>> listOfLists) {
        return listOfLists.stream().flatMap(List::stream).collect(Collectors.toList());
    }

    static <T> List<T> filter(List<T> list, Predicate<T> predicate) {
        return list.stream().filter(predicate).toList();
    }

    static <T> int size(List<T> list) {
        return list.size();
    }

    static <T, U> List<U> map(List<T> list, Function<T, U> transform) {
        return list.stream().map(transform).toList();
    }

    static <T> List<T> reverse(List<T> list) {
        return IntStream.range(0, list.size()).map(i -> list.size() - 1 - i).mapToObj(list::get).toList();
    }

    static <T, U> U foldLeft(final List<T> list, final U initial, final BiFunction<U, T, U> f) {
        if (list.isEmpty()) {
            return initial;
        }

        return foldLeft(new ArrayList<>(list.subList(1, list.size())), f.apply(initial, list.get(0)), f);
    }

    static <T, U> U foldRight(final List<T> list, final U initial, final BiFunction<T, U, U> f) {
        if (list.isEmpty()) {
            return initial;
        }

        return f.apply(list.get(0), foldRight(new ArrayList<>(list.subList(1, list.size())), initial, f));
    }

    private ListOps() {
        // No instances.
    }

}

File: solutions/go/list-ops/1/list_ops.go
package listops

// IntList is an abstraction of a list of integers which we can define methods on
type IntList []int

func (s IntList) Foldl(fn func(int, int) int, initial int) int {
	if len(s) == 0 {
		return initial
	}
	x, xs := s[0], s[1:]
	return xs.Foldl(fn, fn(initial, x))
}

func (s IntList) Foldr(fn func(int, int) int, initial int) int {
	reversedFunc := func(x, y int) int { return fn(y, x) }
	return s.Reverse().Foldl(reversedFunc, initial)
}

func (s IntList) Filter(fn func(int) bool) IntList {
	var result = []int{}
	for _, v := range s {
		if fn(v) {
			result = append(result, v)
		}
	}
	return result
}

func (s IntList) Length() int {
	return len(s)
}

func (s IntList) Map(fn func(int) int) IntList {
	var result = make([]int, s.Length())
	for i, v := range s {
		result[i] = fn(v)
	}
	return result
}

func (s IntList) Reverse() IntList {
	var result = make([]int, s.Length())
	for i, j := 0, s.Length()-1; i < j; i, j = i+1, j-1 {
		result[i], result[j] = s[j], s[i]
	}
	return result
}

func (s IntList) Append(lst IntList) IntList {
	return append(s, lst...)
}

func (s IntList) Concat(lists []IntList) IntList {

	var result = make([]int, 0, s.Length())

	result = append(result, s...)

	for _, l := range lists {
		if l.Length() > 0 {
			result = append(result, l...)
		}
	}
	return result
}
